### PR TITLE
doc: update MetricsBag

### PR DESCRIPTION
**Pull request by lucarin91**
#### :question: What

Update the documentation of the `MetricsBag`.

#### :white_check_mark: Checklists

- [x] Has `make gofmt-write` been run on the code?
- [x] Has `make govet` been run on the code? Has the code been fixed accordingly to the output?
- [ ] Have the changes been added to the [CHANGELOG.md](./CHANGELOG.md) file?
- [x] Have the steps in [CONTRIBUTING.md](./CONTRIBUTING.md) been followed to update a Go module?

### CODE REVIEW

File: metrics.go [hunk 1]
```diff
@@ -11,27 +11,35 @@ import (
 // with values relative to that last second.
 type MetricsBag map[string]interface{}
 
-// AddRawCounter adds a counter that always increments.
+// AddRawCounter adds a counter. A counter is a cumulative metrics that can only increase.
```
Comment (tommyblue) [inline metrics.go:14]:
```suggestion
// AddRawCounter adds a counter. A counter is a cumulative metric that can only increase.
```

File: metrics.go [hunk 2]
```diff
@@ -11,27 +11,35 @@ import (
...
+// value must be positive.
```
Comment (tommyblue) [inline metrics.go:15]: it seems we enforce a check, but we don't, so maybe I'd write:

```suggestion
// To be meaningful, `value` must be positive.
```

File: metrics.go [hunk 3]
```diff
@@ -11,27 +11,35 @@ import (
...
 func (bag MetricsBag) AddRawCounter(name string, value int64) {
 	bag["c:"+name] = value
 }
 
-// AddDeltaCounter adds a count of something that happened in the last second
+// AddDeltaCounter adds a new delta increment for a counter. The increment represents
+// the variation of the counter in the current time interval.
+// delta must be positive.
```
Comment (tommyblue) [inline metrics.go:22]: same as above

File: metrics.go [hunk 4]
```diff
@@ -11,27 +11,35 @@ import (
...
 
-// AddGauge takes a snapshot of a value.
+// AddGauge adds a new snapshot of a value. A Gauge represents a single numerical
+// data point that can arbitrarily go up and down.
 func (bag MetricsBag) AddGauge(name string, value float64) {
 	bag["g:"+name] = value
 }
 
 // AddHistogram adds a set of values to track their statistical distribution.
+// A histogram samples observations and counts them in different 'buckets' in order
```
Comment (tommyblue) [inline metrics.go:34]:
```suggestion
// A histogram sample observes and counts them in different 'buckets' in order
```
Not sure I fully understand what you mean and thus if my suggestion is correct

File: metrics.go [hunk 5]
```diff
@@ -11,27 +11,35 @@ import (
...
+// to track and show the statistical distribution of a set of values.
 func (bag MetricsBag) AddHistogram(name string, values []float64) {
 	bag["h:"+name] = values
 }
 
 // AddTimings adds a set of timings to track their statistical distribution.
+// A histogram samples observations and counts them in different 'buckets'. Timings
```
Comment (tommyblue) [inline metrics.go:41]: same as above

**Review by tommyblue — APPROVED** [review]
lgtm, check the language suggestions

File: metrics.go (hunk 4 above)
Comment (lucarin91) [reply metrics.go:34]: `sample` is the verb and `observation` the noun. Thus, I think that was correct.

**Review by arl — CHANGES_REQUESTED** [review]
I'd expect the doc strings for `AddRawCounter` and `AddDeltaCounter` to be very similar. They both act on the same underlying metric of type **counter** so I'd expect both to hold the definition of a counter, while making it easy to understand their differences.

> A counter is a cumulative metric that can only increase.

Example:
```
AddRawCounter adds a new metric of type counter to the bag and sets its current value.
A counter is a cumulative metric that can only increase. 
```

```
AddDeltaCounter adds a new metric of type counter to the bag and increments its value.
A counter is a cumulative metric that can only increase. 
```

Also I'd add to the other metrics _to the bag_ or _into the bag_. That may seems obvious when you know it but I'd prefer this to be explicit, so it shows that a bag is just a container for metrics, it just stores them but does not publish them in any way.

Last thing, we should improve the documentation of the `Merge` method, explaining how Merge aggregates metrics with the same name coming from 2 different bags.

File: metrics.go [hunk 6]
```diff
@@ -11,32 +11,42 @@ import (
 // with values relative to that last second.
 type MetricsBag map[string]interface{}
 
-// AddRawCounter adds a counter that always increments.
+// AddRawCounter adds a metric of type counter to the bag and sets its current value.
+// A counter is a cumulative metric that can only increase.
+// To be meaningful, `value` must be positive.
```
Comment (arl) [inline metrics.go:16]: In general we don't use backticks in comments as anyway they're not rendered in GoDoc.
So both of the following suggestions are ok (I personally prefer the first one):
```suggestion
// To be meaningful, value must be positive.
```
```suggestion
// To be meaningful, 'value' must be positive.
```

File: metrics.go [hunk 7]
```diff
@@ -11,32 +11,42 @@ import (
...
 func (bag MetricsBag) AddRawCounter(name string, value int64) {
 	bag["c:"+name] = value
 }
 
-// AddDeltaCounter adds a count of something that happened in the last second
+// AddDeltaCounter adds a metric of type counter to the bag and increments its value.
+// A counter is a cumulative metric that can only increase.
+// To be meaningful, `delta` must be positive.
```
Comment (arl) [inline metrics.go:23]: same as above

File: metrics.go [hunk 8]
```diff
@@ -11,32 +11,42 @@ import (
...
-// AddHistogram adds a set of values to track their statistical distribution.
+// AddHistogram adds a metrics of type histogram to the bag and tracks its value.
+// A histogram samples observations and counts them in different 'buckets' in order
+// to track and show the statistical distribution of a set of values.
 func (bag MetricsBag) AddHistogram(name string, values []float64) {
 	bag["h:"+name] = values
 }
 
-// AddTimings adds a set of timings to track their statistical distribution.
+// AddTimings adds a metric of type time to the bag and traks its value.
```
Comment (arl) [inline metrics.go:41]: under the hood, there's no metric of  `time` or `timing`, this is just an wrapper around an histogram tracking time duration, such as latency
```suggestion
// AddTimings adds a metric of type histogram to the bag and tracks its value.
```

File: metrics.go [hunk 9]
```diff
@@ -11,32 +11,42 @@ import (
...
+// Timing is basically a histogram but allows to sample values of type time.Duration.
+// A histogram samples observations and counts them in different 'buckets'.
 func (bag MetricsBag) AddTimings(name string, values []time.Duration) {
 	bag["t:"+name] = values
 }
 
-// Merge merges another MetricsBag into this 'bag'.
+// Merge merges metrics with the same name and types from another MetricsBag into this bag.
+// Counters and Deltas are summed up, Gauge is averaged, and Histograms and Timings concatenate.
```
Comment (arl) [inline metrics.go:49]:
```suggestion
// Counters are summed up, gauges are averaged, and histograms are concatenated.
```

**Review by arl — APPROVED** [review]
LGTM
This was the last pass, there are just some small stuff, you can merge once addressed, thanks!

**Pull request merged**